### DISCUSSION TITLE
How pre- are pre-render scripts?

### DISCUSSION THREAD

**Discussion by andrewpbray** [quarto-dev/quarto-cli#3879] [Q&A] [answered]
When you run `quarto render`, what operations are done before the `pre-render` script is run? I had a pre-render script that changed file paths but on render it errored out because it couldn't find the files at the original paths. That suggests it compiles the list of files to render before running the pre-render script - is that correct?

**Comment by mcanouil** [comment]
You need to actually write something, as is it is merely a post and mostly a spam ...

**Comment by cscheid** [comment, marked as answer]
I agree with @mcanouil that a bit more information would be helpful :) but I think I can attempt to reverse-engineer the question as "when exactly do quarto pre-render scripts get called?"

The short answer to that is "after determining which files are to be rendered, but before rendering any of them". The long answer to that is "the full story is on `src/command/render/project.ts:renderProject`"

**Reply by mcanouil** [reply]
I noticed this order, on my website 😪
I have a pre-render that creates a listing, but the first rendering failed because Quarto detected/checked before the file was created.

@cscheid Do you think the pre-render could be before the checks for the files to be rendered? 🤔

**Reply by cscheid** [reply]
Maybe. This kind of ordering issue is finicky so I don't want to make a change without talking to the rest of the team. I'll get back to you.

**Reply by mcanouil** [reply]
Of course 😉

**Reply by andrewpbray** [reply]
Indeed that was the gist of the question. I convinced myself that was the case - that the list of files to render was built before the pre-render script - and was going to re-write this post addressing a different issue but I couldn't figure out for the life of me how to delete this one. Sorry! I'll go ahead and edit it now to reference render order so that this is a coherent thread.

@mcanouil He's a workaround that I put together to fiddle directly with the list of files to render:

[https://github.com/andrewpbray/scheduled-posts-example](https://github.com/andrewpbray/scheduled-posts-example)

It came about in the context of this [discussion](https://github.com/quarto-dev/quarto-cli/discussions/3114).

**Reply by andrewpbray** [reply]
One consideration I was just thinking about:

If the pre-render script is moved before the render list gets built, it'd be helpful if the latter didn't overwrite existing environmental variables. That would allow us to continue using the pre-render script to build the render list. I'm not sure how one would tinker with the render list otherwise.

**Reply by Nenuial** [reply]
> I noticed this order, on my website 😪
> I have a pre-render that creates a listing, but the first rendering failed because Quarto detected/checked before the file was created.

I have encountered similar issues when using pre-render to setup files on a website for rendering. Two runs of render solve it, but a _better_ way would actually be great!

**Reply by osaal** [reply]
> Maybe. This kind of ordering issue is finicky so I don't want to make a change without talking to the rest of the team. I'll get back to you.

Sorry to return to an old discussion, but has this been considered by the team for a future update?

I have a similar issue, where I wish to change the project file structure before render (as a solution to having source files in their own subdirectory structure, but rendering a website project output that uses a "flattened" structure without the subdirectories). Since the pre-render scripts are ran after finding the files to render (as seen in `src/command/render/project.ts:renderProject`), my current solution is to manually run a script prior to calling `quarto render`. 

Would it be possible to implement a way to run pre-render scripts before Quarto finds the project files to be rendered, or alternatively to suppress erroring on "File not found" in the pre-pre-render part of the render process?

Thanks in advance!
Oliver S.

**Reply by mcanouil** [reply]
This thread is outdated.
What version of Quarto are you using as this should not even be an issue since quite some time now.

**Reply by osaal** [reply]
> This thread is outdated. What version of Quarto are you using as this should not even be an issue since quite some time now.

My apologies. I'm using version 1.6.39. I'll make a separate issue about the problem!

**Reply by mcanouil** [reply]
To be clear, pre-render script can create new document that are discovered by Quarto which was not possible at the time of this thread.
Quarto now runs a second discovery after pre-render.

If you don't see this, please open a separate GitHub discussion with a small reproducible example as described in the guidelines when opening a GitHub Discussion.

**Comment by baptiste** [comment]
Another use-case scenario I've run into: I have a R script that generates `.qmd` files programmatically (each qmd corresponds to a different photo album page for a website); I'm currently running the R script manually before calling `quarto render`, but ideally this would be taken care of by a pre-render step. At the moment the newly-created `.qmd` files would be ignored by quarto.

**Reply by mcanouil** [reply]
Exactly what I described and why I asked if the pre-render could be before file checks (see <https://github.com/quarto-dev/quarto-cli/discussions/3879#discussioncomment-4635259>). 😉
